[WFLY-13866]: Several traces created in Jaeger for just one method call.

ehsavoie opened the pull request:
* the current scope is 'lost' on async calls. Ficxing that by passing
  the scope manager on the new threads.

Jira: https://issues.redhat.com/browse/WFLY-13866

rhusar commented on microprofile/opentracing-extension/src/test/java/org/wildfly/extension/microprofile/opentracing/Subsystem_2_0_ParsingTestCase.java:
```diff
@@ -152,6 +152,7 @@ protected void initializeExtraSubystemsAndModel(ExtensionRegistry extensionRegis
             super.initializeExtraSubystemsAndModel(extensionRegistry, rootResource, rootRegistration, capabilityRegistry);
             Map<String, Class> capabilities = new HashMap<>();
             capabilities.put(WELD_CAPABILITY_NAME, WeldCapability.class);
+            capabilities.put("org.wildfly.microprofile.config", Void.class);
```
Please use org.wildfly.extension.microprofile.opentracing.SubsystemDefinition#MICROPROFILE_CONFIG_CAPABILITY_NAME instead and make it public.

rhusar commented on microprofile/opentracing-extension/src/test/java/org/wildfly/extension/microprofile/opentracing/Subsystem_3_0_ParsingTestCase.java:
```diff
@@ -164,6 +164,7 @@ protected void initializeExtraSubystemsAndModel(ExtensionRegistry extensionRegis
             super.initializeExtraSubystemsAndModel(extensionRegistry, rootResource, rootRegistration, capabilityRegistry);
             Map<String, Class> capabilities = new HashMap<>();
             capabilities.put(WELD_CAPABILITY_NAME, WeldCapability.class);
+            capabilities.put("org.wildfly.microprofile.config", Void.class);
```
Please use org.wildfly.extension.microprofile.opentracing.SubsystemDefinition#MICROPROFILE_CONFIG_CAPABILITY_NAME instead and make it public.

rhusar commented on testsuite/integration/microprofile/src/test/java/org/wildfly/test/integration/microprofile/faulttolerance/opentracing/OpenTracingFaultToleranceTestCase.java:
```diff
@@ -0,0 +1,180 @@
+                Assert.assertEquals(object.toString(), parentId, Integer.parseInt(object.getString("parentId")));
+                switch (i) {
+                    case 0: {
+                        Assert.assertEquals(object.toString(), operationName, "GET:org.wildfly.test.integration.microprofile.faulttolerance.opentracing.application.HelloResource.get");
```
Please future-proof this by using `Java.lang.Class.getCanonicalName()` instead of string constants that will inevitably break with next refactoring.

Ladicek commented on microprofile/fault-tolerance-smallrye/executor/src/main/java/org/wildfly/extension/microprofile/faulttolerance/MiniContextPropagation.java:
```diff
@@ -0,0 +1,285 @@
+
+/**
+ * Miniature context propagation API, adapted from MicroProfile Context Propagation and SmallRye Context Propagation.
+ * @author Emmanuel Hugonnet (c) 2020 Red Hat, Inc.
```
You stole my code! :-D

Really, I'd suggest to:
- make this class package-private
- add a big TODO that this should be replaced by Context Propagation proper when it's integrated in WildFly

rhusar replied:
Certainly a note can be added, but it's not that straightforward – even though if Context Propagation is integrated in WildFly it doesn't mean that the users must provision it in theirs server while still anticipating that the tracing context propagation works - which is what we are doing in the PR.

Ladicek replied:
Yea, the thing is, what this PR does (which is the same thing I did in Thorntail) won't work with other things that also use context propagation. Thorntail is dead, but WildFly isn't, so it deserves better treatment. I'd suggest that if users provision Fault Tolerance and OpenTracing, but don't provision Context Propagation, then they should _NOT_ expect traces being propagated across FT `@Asynchronous` calls. Obviously, that becomes a breaking change if this PR is merged, so...

Again, I could do this hack in Thorntail, because it's dead, but you need to be more careful with WildFly.

ehsavoie replied:
@Ladicek  You mean I took responsibility of this piece of code in WildFly ? ;)

ehsavoie replied:
@Ladicek I disagree with your POV on `@Asynchronous` having tracer spans coming out of nowhere doesn't seem something that make sense. The fact that this issue arose is proof of that to me. If context propagation is provided we should rely on it insted of the current mechanism but otherwise I think we should still provide it.

Ladicek replied:
I agree with you, but the same argument applies to all cases where context propagation is / can be  used, such as RestClient + OpenTracing, @Asynchronous + @RequestScoped, etc. etc. etc. It doesn't make any sense to build a special case support for everything.

On the other hand, of course, Context Propagation is not in MicroProfile yet, so this is all a gray area.

bstansberry replied:
@rhusar WDYT here?

rhusar replied:
I am on the fence here, spent a lot of time thinking about this, here is some summary:

* All of this is out of the spec – there is no mention in FT nor OpenTracing.
* We can't leverage Context Propagation for this as its not part of the platform yet – we will when it does make it to the platform.
* If we bend over backwards now to support this use case, going forward, we will have to deal with complexity once Context Propagation is available.
* Yet, it does make sense to propagate the context in this scenario, but if we are doing that, we should evaluate all cases across the board, rather than a single one.
* The original issue was reported by QE, not a real world user use case; so e.g. introducing this in XP2 is a behavior change which might not be expected.

I think it boils down to what users are expecting to happen and @Ladicek 's comment "I'd suggest that if users provision Fault Tolerance and OpenTracing, but don't provision Context Propagation, then they should NOT expect traces being propagated across FT @Asynchronous calls." suggests that we shouldn't move forward with this.

rhusar commented on testsuite/integration/microprofile/src/test/java/org/wildfly/test/integration/microprofile/faulttolerance/opentracing/application/MyService.java:
```diff
@@ -0,0 +1,65 @@
+            throw new RuntimeException("failed");
+        }
+
+//        tracer.buildSpan("hello").start();
+//        tracer.scopeManager().activeSpan().log("attempt " + counter.getAndIncrement());
+//        throw new RuntimeException("failed");
```
Maybe this can be gone?

bstansberry commented on microprofile/fault-tolerance-smallrye/executor/src/main/java/org/wildfly/extension/microprofile/faulttolerance/MiniContextPropagation.java:
```diff
@@ -0,0 +1,286 @@
+ * @author Ladislav Thon (c) 2020 Red Hat, Inc.
+ * @author Emmanuel Hugonnet (c) 2020 Red Hat, Inc.
+ */
+public final class MiniContextPropagation {
```
If this and other new classes can be package private that is better.

It's a private module but locking things down is better.

ehsavoie committed:
[WFLY-13866]: Several traces created in Jaeger for just one method call.

* the current scope is 'lost' on async calls. Ficxing that by passing
  the scope manager on the new threads.

Jira: https://issues.redhat.com/browse/WFLY-13866

bstansberry commented:
Per the discussion starting at https://github.com/wildfly/wildfly/pull/13590#discussion_r492612456 we've decided not to proceed with this approach and we'll defer dealing with WFLY-13866 issue until we have a more general idea of how MP Context Propagation will integrate (e.g. https://issues.redhat.com/browse/WFLY-13921), or at least a broader review of places where the same kind of thing would be needed.
